### PR TITLE
Week 6: Implement calendar that can schedule meetings among multiple people.

### PR DESCRIPTION
**Pull request by lilymzhou**
- Supports basic functionality from existing tests
  - accounts for: no attendees, impossible request, impossible schedules, overlapping events, nested events, etc.
- Supports additional tests described in walkthrough
  - account for optional attendees: if one or more time slots exists so that both mandatory and optional attendees can attend, return those time slots. Otherwise, return the time slots that fit just the mandatory attendees.

### CODE REVIEW

**Comment by lilymzhou** [comment]
cc: @tjgoog

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,156 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Add a javadoc comment explaining what the query method does.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,156 @@
...
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> slots = new ArrayList<TimeRange>();
+
+    ArrayList<Event> eventsArr = sortEventsByStart(events);
+
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return slots;
+    }
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: It's generally more readable to check invalid conditions at the very top of the method and immediately return before creating any local variables.

I would also add a comment to explain what this check is and why requests for meetings longer than a day are invalid.

```suggestion
    // Request duration is longer than a day, don't return anything (why?).
    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
      return new ArrayList<TimeRange>();
    }
    
    ArrayList<TimeRange> slots = new ArrayList<TimeRange>();
    ArrayList<Event> eventsArr = sortEventsByStart(events);    
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,156 @@
...
+
+    slots.add(TimeRange.WHOLE_DAY);
+
+    if (!request.getOptionalAttendees().isEmpty()) {
+      ArrayList<TimeRange> optionalSlots = findTimeSlots(eventsArr, request, slots, true);
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: When a method has multiple parameters like this, it can be hard to keep track of what each param means especially if it has a value like true or false.  In this case, we can leave a comment to better document what this arg means:
```suggestion
      ArrayList<TimeRange> optionalSlots = findTimeSlots(eventsArr, request, slots, /* hasOptionalAttendees= */ true);
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,156 @@
...
+
+    ArrayList<Event> eventsArr = sortEventsByStart(events);
+
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return slots;
+    }
+
+    slots.add(TimeRange.WHOLE_DAY);
+
+    if (!request.getOptionalAttendees().isEmpty()) {
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: you could also add a method to MeetingRequest for hasOptionalAttendees() which could do this for you

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,156 @@
...
+    });
+    return eventsArr;
+  }
+
+  /*
+   * Returns a list with all the available time slots when the requested meeting
+   * can take place.
+   */
+  private ArrayList<TimeRange> findTimeSlots(Collection<Event> events, MeetingRequest request,
+    ArrayList<TimeRange> slots, boolean hasOptionalAttendees) {
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: i'm having trouble understanding what slots and hasOptionalAttendees mean, it looks like slots just has TimeRange.WHOLE_DAY when this is first called?  Does hasOptionalAttendees mean findTimeSlots should take into account optional attendees and only return time slots that work with those optional attendees?

It doesn't look like this is ever called without slots equaling WHOLE_DAY so can this just be a local variable?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,156 @@
...
+      TimeRange eventTime = event.getWhen();
+      slots = curSlots;
+
+      // Determine whether event's attendees overlaps with request's attendees.
+      if (hasOptionalAttendees && !hasAllAttendees(event, request)) {
+        continue;
+      }
+      if (!hasOptionalAttendees && !hasRequiredAttendees(event, request)) {
+        continue;
+      }
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:103]: So here you're trying to determine if this even should "disqualify" the TimeRange that this event takes place during.  If either of these checks are true, the event is not a problem and shouldn't be considered.  Otherwise you have to figure out how to carve out the event's time slot from your running list of options.

I think this can be documented better via comments and also variable/method names since I found it to be a little difficult to follow this code.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,156 @@
...
+   * Removes everything in the slot after TimeRange before, corresponding to the space
+   * that the event takes (and thus the meeting cnanot take place in).
+   */
+  private void removeFromStart(ArrayList<TimeRange> curSlots, long reqDuration,
+    TimeRange before, TimeRange slot) {
+    if (before.duration() >= reqDuration) {
+      curSlots.add(before);
+    }
+    curSlots.remove(slot);
+  }
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: These three methods look similar, with removeFromMiddle essentially being a combination of removeFromStart and removeFromEnd.  Can they be combined?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,156 @@
...
+
+  /*
+   * Return whether attendees for event overlap with required attendees for request
+   * (i.e. at least one attendee for request is an attendee for event).
+   */
+  private boolean hasRequiredAttendees(Event event, MeetingRequest request) {
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This method is only ever used with ! hasRequiredAttendees so perhaps it's more understandable if this determined if required attendees were *not* present during this event. 

I think it's especially confusing because hasRequiredAttendees makes it seems like this will return true if the event contains all the request's required attendees, which is not what this is doing.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -14,10 +14,156 @@
...
+    }
+    return false;
+  }
+
+  /*
+   * Return whether attendees for event overlap with all (required 
+   * and optional) attendees for request (i.e. at least one attendee 
+   * for request is an attendee for event).
+   */
+  private boolean hasAllAttendees(Event event, MeetingRequest request) {
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: similar comment as above - perhaps you can just check the opposite and/or update the naming to better reflect this logic.

**Review by megankj — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (tjgoog) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: When I write a method's Javadoc comment, I try to make sure that each parameter is mentioned in the comment in a way that describes its meaning or use, or if I think its meaning is super obvious, I double-check that belief before leaving it out. I think that a Javadoc that refers to all of these parameters could help explain how the pieces fit together.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (tjgoog) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: And for this parameter, if I'm understanding its meaning correctly, it's not exactly about whether the request has optional attendees, but whether to only return time slots that the optional attendees can attend. So `treatOptionalAttendeesAsRequired` or `mustIncludeOptionalAttendees` or something would be better IMO.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 9 above)
Comment (tjgoog) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Also, optionally, consider replacing these with a method that checks whether there's any overlap in two groups of attendees (represented by Collections of Strings) and pass different groups of attendees to it for the different checks. e.g.
```
private static boolean hasAnyInCommon(Collection<String> groupA, Collection<String> groupB) {
  // TODO
}

// inside findTimeslots()...
if (hasOptionalAttendees && !hasAnyInCommon(event.getAttendees(), request.getAttendees()) && !hasAnyInCommon(event.getAttendees(), request.getOptionalAttendees()) {
  continue;
}
if (!hasOptionalAttendees && !hasAnyInCommon(event.getAttendees(), request.getAttendees())) {
  continue;
}
// ...end excerpt from findTimeslots()

// or, another way to do the logic in findTimeslots()...
if (!hasAnyInCommon(event.getAttendees(), request.getAttendees()) {
  if (!hasOptionalAttendees || !hasAnyInCommon(event.getAttendees(), request.getOptionalAttendees()) {
    continue;
  }
}
// ...end excerpt from findTimeslots()

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (lilymzhou) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: ah I didn't realize that slots is always the same when findTimeSlots() is called -- I'm moving the declaration of slots within findTimeSlots and renaming hasOptionalAttendees to includeOptionalAttendees.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 8 above)
Comment (lilymzhou) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I changed the method names to `overlapWithRequiredAndOptionalAttendees` and `overlapWithRequiredAttendees`. I found that `overlapWithRequiredAttendees` is used without ! in a later function, so I ultimately kept both methods in the positive form in order to avoid using a double negative (i.e. `!noOverlapWithRequiredAttendees`).

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 9 above)
Comment (lilymzhou) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: thanks so much -- I think this would be a lot cleaner/more intuitive to read.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -14,10 +14,169 @@
...
+      /*
+       * Finds whether there is any overlap between event's attendees and request's attendees.
+       * If there is no overlap, then this event is skipped.
+       */
+      if (hasNothingInCommon(event.getAttendees(), request.getAttendees())) {
+        if (!includeOptionalAttendees || 
+          hasNothingInCommon(event.getAttendees(), request.getOptionalAttendees())) {
+          continue;
+        }
+      }
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: The spacing in this suggestion may be a little off, but this can be combined into one:
```suggestion
      if ((hasNothingInCommon(event.getAttendees(), request.getAttendees()))) &&
             (!includeOptionalAttendees ||  
             hasNothingInCommon(event.getAttendees(), request.getOptionalAttendees()))){
          continue;
      }
```

Also, I love this change!  I think it's much easier to read now.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -14,10 +14,169 @@
...
+    }
+    curSlots.remove(slot);
+  }
+
+  /*
+   * @groupA, groupB: generalized lists of strings to be compared.
+   * @return whether there is any overlap between groupA and groupB (i.e. a String
+   * appears in both groupA and groupB.
+   */
+  private boolean hasNothingInCommon(Collection<String> groupA, Collection<String> groupB) {
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: So this is basically checking if the intersection of two sets is empty.  You could look into using retainAll instead of doing this yourself.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 12]
```diff
@@ -270,5 +272,193 @@ public void notEnoughRoom() {
...
+        new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_C);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+            TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+            TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:297]: General comment: A good way to structure tests is in three distinct steps - arrange, act, and assert.

- arrange: setting up details of the external world specific to the situation under test (creating local variables, etc)
- act: make the call to the method you are testing
- assert: make assertions

If they're in three distinct visual blocks like this, it's easier for people to understand how the test is working.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 13]
```diff
@@ -270,5 +272,193 @@ public void notEnoughRoom() {
...
+            Arrays.asList(PERSON_B))
+    );
+
+    MeetingRequest request =
+        new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_C);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: since you have such a great little diagram above, it might be easier to give each of these TimeRange variables a name like 
`TimeRange option1 = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false);`

So it's easy to see how it corresponds to your diagram above!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 14]
```diff
@@ -270,5 +272,193 @@ public void notEnoughRoom() {
...
+  @Test
+  public void justEnoughRoomWithOptionalAttendee() {
+    // Have one required attendee A, such that there is just enough room 
+    // at one point in the day to have the meeting. B is added an optional
+    // attendee, but it is impossible to schedule a meeting with B; thus,
+    // B is ignored.
+    //
+    // Events  : |--A--|-B-| |----A----|
+    // Day     : |---------------------|
+    // Options :
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: should this have an option in this diagram?

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 15]
```diff
@@ -270,5 +272,193 @@ public void notEnoughRoom() {
...
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+            TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+            TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void optionalAttendeesWithNoGaps() {
+    // Have each person have different events. We should see two options because each person has
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: we should see no options, not two, right?

**Review by megankj — CHANGES_REQUESTED** [review]
This is looking good, I think you've made some excellent changes for readability and comprehension.  Thanks for all the documentation you've left in the code!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 11 above)
Comment (lilymzhou) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: It seems like using retainAll would modify the Collection objects themselves -- I replaced this method with a call to Collections.disjoint() instead.

**Review by megankj — APPROVED** [review]
Please allow Tim to review before merging.  Thanks!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 16]
```diff
@@ -14,10 +14,154 @@
...
+   * @return a list of the possible time ranges for the meeting request to take place.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    /*
+     * Request duration is longer than a day, which is impossible to find an appropriate
+     * time range for. Returns an empty list.
+     */
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return new ArrayList<TimeRange>();
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I think you can leave out the explicit type parameter and do `new ArrayList<>()` and the compiler should infer it. Or you can do `Collections.emptyList()` as another way to boil this down further.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 17]
```diff
@@ -14,10 +14,154 @@
...
+    }
+    
+    ArrayList<Event> eventsArr = sortEventsByStart(events);
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I think this should be called `sortedEvents` to differentiate it from `events`.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 18]
```diff
@@ -14,10 +14,154 @@
...
+
+    if (request.hasOptionalAttendees()) {
+      ArrayList<TimeRange> optionalSlots = findTimeSlots(eventsArr, request, 
+        /* includeOptionalAttendees = */ true);
+      if (optionalSlots.isEmpty() && request.hasAttendees()) {
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Rather than repeat the same call a few lines later, I think it would be cleaner to organize the logic so that both the "has optional attendees but they can't make it" and the "doesn't have optional attendees" cases fall through to the same code. e.g.
```
if (request.hasOptionalAttendees()) {
  ArrayList<TimeRange> optionalSlots = find....;
  if (!optionalSlots.isEmpty(){
    return optionalSlots;
  }
}
return findTimeSlots....;
```
Your call if you want to do (or something entirely different) based on whatever you think is the easiest to read and maintain.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 19]
```diff
@@ -14,10 +14,154 @@
...
+      }
+    });
+    return eventsArr;
+  }
+
+  /*
+   * Finds all the possible time ranges throughout the day that the requested
+   * meeting can take place, given already-scheduled events.
+   * 
+   * @param events: List of already-scheduled events that may conflict with the times
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:75]: Does `events` need to be already sorted by start time? If so, that should be documented and the parameter type should be a `List<Event>` because not all `Collection`s are guaranteed to have a specific order, but `List`s are. If not, might as well remove the sorting in `query()`, right?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 20]
```diff
@@ -14,10 +14,154 @@
...
+
+    /*
+     * A copy of slots that stores new changes to slots without interferring
+     * with the for loop below.
+     */
+    ArrayList<TimeRange> curSlots = new ArrayList<>(slots);
+
+    for (Event event : events) {
+      TimeRange eventTime = event.getWhen();
+      slots = curSlots;
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Variables in Java for objects (i.e. any value that's not one of the primitive types like `int`, `boolean`, etc) are always references (analogous to pointers in C/C++) to the actual object. Because of that, this assignment doesn't make a copy of the list, it just sets `slots` to point to the same list that `curSlots` does, so you should be able to simplify the code down to only using one of the two list variables.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 21]
```diff
@@ -14,10 +14,154 @@
...
+          /* addBeginningSlot = */ false, /* addEndSlot = */ false);
+        }
+      }
+    }
+    return curSlots;
+  }
+
+  /*
+   * Split current time slot into 0-2 smaller time slots, given that event at least 
+   * partially overlaps slot. If event is a subset of slot, then both the remaining
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Tend toward using `{@code slot}` (or `{@code event}` or whatever the parameter name is) when referring to the name of a parameter in a sentence in order to emphasize that this is the name of a specific thing rather than the general concept of an event or slot.

When run through a tool that interprets Javadoc (e.g. some IDEs, or the Javadoc html generator thing), the tool will style the contained text differently, usually bold and/or with a fixed width font. In the source files, it's not nearly as pretty, but it stands out and gets the point across.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 22]
```diff
@@ -14,10 +14,154 @@
...
+   * @param eventTime: time range that the event currently be considered takes place in.
+   * @param addBeginningSlot: whether the time range (beginning of slot - beginning of
+   * event) should be considered as a valid time slot for the requested meeting.
+   * @param addEndSlot: whether the time range (end of event - end of slot) should
+   * be considered as a valid time slot for the requested meeting.
+   */
+  private void splitSlotAroundEvent(ArrayList<TimeRange> curSlots, long reqDuration,
+    TimeRange slot, TimeRange eventTime, boolean addBeginningSlot, boolean addEndSlot) {
+
+    TimeRange before = TimeRange.fromStartEnd(slot.start(), eventTime.start(), false);
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Could you add `/* inclusive= */` before these boolean values?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 23]
```diff
@@ -14,10 +14,154 @@
...
+    TimeRange after = TimeRange.fromStartEnd(eventTime.end(), slot.end() - 1, true);
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Does it work to set the last param to false instead and then avoid subtracting 1 from the new TimeRange's ending value?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 24]
```diff
@@ -14,10 +14,154 @@
...
+       * If there is no overlap, then this event is skipped.
+       */
+      if ((Collections.disjoint(event.getAttendees(), request.getAttendees())) &&
+        (!includeOptionalAttendees ||  
+        Collections.disjoint(event.getAttendees(), request.getOptionalAttendees()))){
+        continue;
+      }
+
+      for (TimeRange slot : slots) {
+        if (slot.contains(eventTime)) { // Event is a subset of slot.
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I _think_ it might be possible to collapse all these cases down. It would seem that it's always safe to set `addBeginningSlot` and `addEndSlot` to `true`, assuming that all existing events, slots, and the requested meeting duration are non-negative, because the duration checks in `splitSlotAroundEvent` seem to automatically handle whether or not to add the beginning and/or end.

If that's correct, then you can just remove the two boolean params. And then once they're removed, all of the calls here are identical, so you can just do one call if there's any overlap between the slot and the existing event, based on `TimeRange.overlaps()`.

I'm not sure of my mathematical reasoning here, but if it sounds right to you, might as well try it because tests make it easy to verify.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 25]
```diff
@@ -262,13 +256,192 @@ public void notEnoughRoom() {
...
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartDuration(TIME_0830AM, DURATION_30_MINUTES));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void optionalAttendeesWithGaps() {
+    // Have each person have different events. We should see three options because each person has
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: I think it's worth noting in this description that all of the attendees are optional.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 26]
```diff
@@ -262,13 +256,192 @@ public void notEnoughRoom() {
...
 
     Collection<TimeRange> actual = query.query(events, request);
     Collection<TimeRange> expected = Arrays.asList();
 
     Assert.assertEquals(expected, actual);
   }
-}
 
+  @Test
+  public void impossibleOptionalAttendeeMixedOrder() {
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:417]: Is this different from `impossibleOptionalAttendee()`?

**Review by tjgoog — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 18 above)
Comment (lilymzhou) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: thanks -- this is a lot cleaner!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 19 above)
Comment (lilymzhou) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:75]: It seems like changing the parameter from `Collection<Event> events` to `List<Event> events` causes errors when query() is called, even though ArrayLists are always used to call the function. Is it sufficient to specify in the javadoc that the input should be a `List<Event>`?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 23 above)
Comment (lilymzhou) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: yep, this seems to work!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 24 above)
Comment (lilymzhou) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: yes, this seems to work -- thanks for pointing this out!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 26 above)
Comment (lilymzhou) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:417]: This essentially tests to make sure that query() is working regardless of the order that `events` is in -- within query(), this list is eventually sorted by start order.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 19 above)
Comment (tjgoog) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:75]: Following up from some Chat messages, I think that the best approach is to avoid sorting this input at all, if sorting the output will work instead, assuming that sorted output is the ultimate purpose of sorting the input. Sorting the output would more directly express the intended behavior and be more robust to future changes that might rearrange the slots as they're being checked.

If sorting the output isn't a working/ideal approach, I think that sorting the input at the beginning of this function, rather than relying on the caller to do it before passing it in, would be a good approach. It makes this method more "fool-proof" to use, which becomes more and more important as code is called from more places, especially if people ignore its documentation that says how the input should be organized.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 27]
```diff
@@ -14,10 +14,132 @@
...
+      if ((Collections.disjoint(event.getAttendees(), request.getAttendees())) &&
+        (!includeOptionalAttendees ||  
+        Collections.disjoint(event.getAttendees(), request.getOptionalAttendees()))){
+        continue;
+      }
+
+      for (TimeRange slot : slots) {
+        if (slot.overlaps(eventTime)) {
+          splitSlotAroundEvent(slots, request.getDuration(), slot, eventTime);
+          break;
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I think that this `break` statement could cause incorrect output in some cases that don't currently have tests. Discussed more in Chat about why I think so and what could be done instead. Can continue conversation here or in Chat.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 28]
```diff
@@ -262,13 +256,193 @@ public void notEnoughRoom() {
...
+
+    Collection<TimeRange> actual = query.query(events, request);
+    TimeRange option1 = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false);
+    TimeRange option2 = TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false);
+    TimeRange option3 = TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true);
+    Collection<TimeRange> expected = Arrays.asList(option1, option2, option3);
+
+    Assert.assertEquals(expected, actual);
+  }
+}
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:474]: Here's a test for the corner case that `beak` may be breaking.

```suggestion
 @Test
  public void oneEventSplitsTwoExistingSlots() {
    // First event splits the available all-day slot in two, with a half hour gap, then second event starts
    // before and ends after the gap should affect both of the two slots.
    // 
    //
    // Events  :              |--A--|
    // Events  :          |------B------|
    // Day     : |-------------------------------|
    // Options : |---1---|               |---2---|
  
    Collection<Event> events = Arrays.asList(
        new Event("Event 1", TimeRange.fromStartDuration(TIME_0900AM, DURATION_30_MINUTES),
            Arrays.asList(PERSON_A)),
        new Event("Event 2", TimeRange.fromStartDuration(TIME_0800AM, DURATION_2_HOUR),
            Arrays.asList(PERSON_B)),
    );
    MeetingRequest request =
        new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);

    Collection<TimeRange> actual = query.query(events, request);
    TimeRange option1 = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false);
    TimeRange option2 = TimeRange.fromStartEnd(TIME_1000AM, TimeRange.END_OF_DAY, true);
    Collection<TimeRange> expected = Arrays.asList(option1, option2);

    Assert.assertEquals(expected, actual);
  }
  }
}
```

**Review by tjgoog — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 29]
```diff
@@ -121,25 +121,27 @@ public int compare(Event a, Event b) {
    * of event) or (end of event - end of slot) are at least as long as {@code reqDuration}
    * they are kept in {@code curSlots} as possible meeting times.
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Update this part of the description since switching to newSlots/removeSlots version.

**Review by tjgoog — APPROVED** [review]
Just a tiny documentation update request before merging
